test/test_nestedtensor.py

```diff
@@ -4339,6 +4353,115 @@ def forward(self, query, value, offsets):
         self.assertTrue(torch.allclose(attn_output_eager, attn_output))
         self.assertTrue(torch.allclose(value_grad, value.grad))
 
+    @dtypes(torch.float32)
+    @skipIfTorchDynamo("Test compiles internally")
+    @unittest.skipIf(sys.version_info >= (3, 12), "torch.compile is not supported on python 3.12+")
+    @unittest.skipIf(IS_WINDOWS, reason="Windows not yet supported for torch.compile")
+    @skipCUDAIf(not SM70OrLater, "GPU capability is < SM70")
+    def test_compile_preserves_metadata_cache(self, device, dtype):
+        # shape (B, *, D)
+        nt = random_nt_from_dims(
+            [4, None, 3, 16], device=device, dtype=dtype, layout=torch.jagged, requires_grad=True)
+
+        # expect min / max seqlen to be stored here
+        cache = dict(nt._metadata_cache)
+
+        @torch.compile
+        def f(nt):
+            q = nt.transpose(-3, -2)
+            output = F.scaled_dot_product_attention(q, q, q).transpose(-3, -2)
+            return output
+
+        output = f(nt)
+        output.backward(torch.ones_like(output))
+        self.assertEqual(output._metadata_cache, cache)
+
+    @dtypes(torch.float32)
+    @skipIfTorchDynamo("Test compiles internally")
+    @unittest.skipIf(sys.version_info >= (3, 12), "torch.compile is not supported on python 3.12+")
+    @unittest.skipIf(IS_WINDOWS, reason="Windows not yet supported for torch.compile")
+    @skipCUDAIf(not SM70OrLater, "GPU capability is < SM70")
+    def test_compile_with_dynamic_max_seq_len(self, device, dtype):
+        # shape (B, *, D)
+        # max seq len: 18
+        nt = torch.nested.nested_tensor([
+            torch.randn(2, 5),
+            torch.randn(3, 5),
+            torch.randn(18, 5),
+        ], layout=torch.jagged)
+
+        # max seq len: 19
+        nt2 = torch.nested.nested_tensor([
+            torch.randn(2, 5),
+            torch.randn(3, 5),
+            torch.randn(19, 5),
+        ], layout=torch.jagged)
+
+        def f(nt):
+            return torch.ones_like(nt) * nt.max_seqlen()
+
+        for dynamic in [False, True]:
+            self.assertFalse(_recompiles_for_inputs(f, (nt,), (nt2,), dynamic=dynamic))
```

Comment:
jfyi there's `with unittest.mock.patch("torch._dynamo.config.error_on_recompile", True):` though `_recompiles_for_inputs` seems generally useful if you also want to make sure cases where recompiles=True

Reply:
I lifted `_recompiles_for_inputs()` from some other test file, which isn't great :p